build-system/tasks/e2e/README.md

```diff
@@ -0,0 +1,181 @@
+
+Test Type | Execution time (order of magnitude)
+--|--
+Unit | ~10ms each
+Integration | ~100ms each
+End-to-end | ~1000ms each
+
+
+## Choosing which features to test
+
+End-to-end tests should verify the most important user flows of a component. Prioritize tests for behaviors that would make the page appear very obviously broken if there was a failure. For example:
+
+* The component's initial render
+* Primary user interactions
+  * e.g. clicking the next button on a carousel
+* Features with heavy usage by a large number of AMP publishers
+  * e.g. loading more content at the bottom of an amp page
+* Important behaviors that are frequently broken
+  * e.g. browser updates often break video autoplay behavior
+
+
+## Writing E2E tests
+
+AMP uses an abstraction layer that is implemented using major end-to-end testing frameworks like Selenium. The structure of the tests is similar to existing integration tests for AMP, using `describes.endtoend`.
+
+Create your test JS files by adding them to `test/e2e` or `extensions/**/test-e2e`. For examples, see `extensions/amp-base-carousel/0.1/test-e2e/*.js`.
+
+```js
+// extensions/amp-foo/0.1/test-e2e/test-amp-foo-basic.js
+
+describes.endtoend('amp-foo', {
+  testUrl: 'http://localhost:8000/test/manual/amp-foo/amp-foo-basic.amp.html',
+  experiments: ['amp-foo'],
+  // By default, the browser opens at 800x600
+  // initialRect: {width: 800, height: 600},
+
+  // By default, E2E tests run in all three environments
+  // environments: ['single', 'viewer-demo', 'shadow-demo']
+}, env => {
+  let controller;
+
+  beforeEach(() => {
+    controller = env.controller;
+  });
+
+  it('should render correctly', async() => {
+    const fooBars = await controller.findElements('.i-amphtml-foo-bar');
+    await expect(fooBars).to.have.length(2);
+    await expect(controller.getElementText(fooBars[0])).to.equal('baz');
+
+    /* ... */
+  });
+});
+```
+
+AMP E2E tests have a few differences from AMP integration and unit tests.
+
+1. `async`/`await`
+2. asynchronous `expect`
+3. implicit polling
+
+### async/await
+
+Your end-to-end tests should use `async`/`await` instead of Promises. E2E tests are inherently asynchronous, and using Promises would become verbose quickly.
+
+### asynchronous expect
+
+Another change is that `expect` is now asynchronous, so we always put `await` before calls to `expect`. It is not necessary in 100% of cases, but when the test will break if it is necessary and it is not present, so we prefer to always use `await` before `expect`. This change enables us to use implicit polling.
```

Comment:
How about something like:

> Another change is that `expect` is now asynchronous, so we always put `await` before calls to `expect`. While it is not necessary in all cases, you should always use `await` with `expect` to maintain consistency and to ensure that you do not inadvertently forget to add it. Otherwise, tests will not behave correctly (and may be flaky) when the `expect` is working on an asynchronous operation.